pkg/api/message/subscribe_worker.go

```diff
@@ -22,207 +22,96 @@ const (
-			continue
-		}
-		logger.Debug("adding topic listener", zap.String("topic", validatedTopic.String()))
-		l.topics[validatedTopic.String()] = struct{}{}
-	}
-
-	for _, originator := range originators {
-		l.originators[originator] = struct{}{}
-	}
-
-	return l
-}
-
-type listenerSet struct {
-	sync.Map // map[*listener]struct{}
-}
-
-func (ls *listenerSet) addListener(l *listener) {
-	ls.Store(l, struct{}{})
-}
-
-func (ls *listenerSet) removeListener(l *listener) {
-	ls.Delete(l)
-}
-
-func (ls *listenerSet) isEmpty() bool {
-	empty := true
-	ls.Range(func(_, _ any) bool {
-		empty = false
-		return false // stop iteration
-	})
-	return empty
-}
-
-// Maps from a key to a set of listeners
-type listenersMap[K comparable] struct {
-	data sync.Map     // map[K]*listenerSet
-	mu   sync.RWMutex // ensures mutations are consistent
-}
-
-func (lm *listenersMap[K]) addListener(keys map[K]struct{}, l *listener) {
-	lm.mu.RLock()
-	defer lm.mu.RUnlock()
-	for key := range keys {
-		value, _ := lm.data.LoadOrStore(key, &listenerSet{})
-		set := value.(*listenerSet)
-		set.addListener(l)
-	}
-}
-
-func (lm *listenersMap[K]) removeListener(keys map[K]struct{}, l *listener) {
-	lm.mu.Lock()
-	defer lm.mu.Unlock()
-	for key := range keys {
-		value, ok := lm.data.Load(key)
-		if !ok || value == nil {
-			return // Key doesn't exist, nothing to do
-		}
-		set := value.(*listenerSet)
-		set.removeListener(l)
-		if set.isEmpty() {
-			lm.data.Delete(key)
-		}
-	}
-}
-
-func (lm *listenersMap[K]) getListeners(key K) *listenerSet {
-	// No lock needed, because we are not mutating lm.data
-	if value, ok := lm.data.Load(key); ok {
-		return value.(*listenerSet)
-	}
-	return nil
-}
-
-func (lm *listenersMap[K]) rangeKeys(fn func(key K, listeners *listenerSet) bool) {
-	lm.data.Range(func(key, value any) bool {
-		return fn(key.(K), value.(*listenerSet))
-	})
-}
-
 // A worker that listens for new envelopes in the DB and sends them to subscribers
 // Assumes that there are many listeners - non-blocking updates are sent on buffered channels
 // and may be dropped if full
 type subscribeWorker struct {
 	ctx    context.Context
 	logger *zap.Logger
+	store  *db.Handler
+
+	// Keep track of known originators.
+	registry registry.NodeRegistry
+
+	// Poll envelopes per originator
+	subscriptions *subscriptionHandler
 
-	dbSubscription <-chan []queries.GatewayEnvelopesView
 	// Assumption: listeners cannot be in multiple slices
 	emptyListeners      listenerSet
 	originatorListeners listenersMap[uint32]
 	topicListeners      listenersMap[string]
 }
 
+func (s *subscribeWorker) getOriginatorNodeIds() ([]uint32, error) {
+	// Get initial list of nodes.
+	nodes, err := s.registry.GetNodes()
+	if err != nil {
+		return nil, fmt.Errorf("could not get list of originators: %w", err)
+	}
+
+	var nodeIDs []uint32
+	for _, node := range nodes {
+		// TODO: Check - this seems like a sane thing to do but requires updating tests.
+		// if !node.IsCanonical {
+		// 	continue
+		// }
```

Comment:
This check made sense to me but some tests were failing originally, so I commented it out.

If it indeed makes sense I can go around and update the tests and reinstate it

Reply:
you should never receive non-canonical nodes. What is the failure?

Reply:
Nothing serious, some tests are relaxed and create nodes that are not canonical (and corresponding envelopes). Suddenly subscribe worker is more strict and rejects them, and tests fail.

Just need to go through the tests and update the nodes with the `IsCanonical` flag set to true.